<html>
    <head>
        <title>qPCR Simulation App Demo</title>
        <script language="javascript" src="Hybridizer.js"></script>
        <script language="javascript" src="PrimerSearcher.js"></script>
        <script language="javascript" src="TRANSCRIPT_TEMPLATES.js"></script>  <!-- instantiates global TEMPLATES object -->
        <script language="javascript" src="PATIENT_TEMPLATE_QUANTITIES.js"></script>  <!-- instantiates global PATIENT_TEMPLATE_QUANTITIES object -->
        <script language="javascript" src="Solution.js"></script>
        <script language="javascript" src="PCR.js"></script>
        <script language="javascript" src="gel.js"></script>
        <!-- script language="javascript" src="bob_oligo.js"></script -->
        <!-- script language="javascript" src="CACHE_PRELOADS.js"></script -->
    </head>
    <body>
        <script language="javascript">
            var SEARCHER = new PrimerSearcher();
            var HYBRIDIZER = new Hybridizer();
            var GEL = new Gel(display_div_id="gel_display");

            // SEARCHER.load_cache_from_data_obj(CACHE_PRELOADS)

            // aal = SEARCHER.search_primer('NC_000913.3', 'ATGTTCAGCGTTCTGGTC', strand='top', fudge=10);a = aal[0].alignments[0]

            go = function(){
                document.getElementById("status_div").innerHTML = "<b>Running...</b>"
                window.setTimeout(run_PCR_reactions, 1)
            }

            var GEL_DATA;  // set as a side effect of 'run_PCR_reactions'
            var PCRs = {}
            var lanes_data; // !!! global for debugging !!!

            set_up_PCR_reaction = function(primerA_seq, primerB_seq, ptq, volume=10, num_cycles=25, annealing_temp=55){
                let my_soln = new Solution('soln1', volume=10)
                
                let primerA = new DnaMolecule(primerA_seq, primerA_seq)
                let primerB = new DnaMolecule(primerB_seq, primerB_seq)

                my_soln.add_ingredient(primerA, 1e-6)
                my_soln.add_ingredient(primerB, 1e-6)
                
                for (const [ptq_key, ptq_value] of Object.entries(ptq)){
                    if (ptq_key in TEMPLATES){  // this is a gene name
                        let my_template = new DnaMolecule(ptq_key, TEMPLATES[ptq_key])
                        my_soln.add_ingredient(my_template, ptq_value/1e15)  // !!! adjust as needed!!!
                    }
                }
                
                let my_pcr = new PCR(my_soln)
                
                return(my_pcr)
            }

            run_PCR_reactions = function(){

                let primerA_seq = document.getElementById('primerA').value
                let primerB_seq = document.getElementById('primerB').value
                let num_cycles = document.getElementById('pcr_num_cycles').value
                let annealing_temp = document.getElementById('pcr_annealing_temp').value

                for (let ptq of PATIENT_TEMPLATE_QUANTITIES){
                    my_pcr = set_up_PCR_reaction(
                        primerA_seq, 
                        primerB_seq,
                        ptq,
                        volume=10
                    )
                     
                    my_pcr.run(num_cycles=num_cycles, denaturationTemp=98, annealingTemp=annealing_temp)
                    PCRs[ptq.patient_id] = my_pcr
                }

                lanes_data = [
                    GEL.create_marker_band_data(100, 20),
                    PCRs[1].get_band_data(),
                    PCRs[2].get_band_data(),
                    PCRs[3].get_band_data(),
                    PCRs[4].get_band_data(),
                    PCRs[5].get_band_data(),
                    PCRs[6].get_band_data(),
                    PCRs[7].get_band_data(),
                    PCRs[8].get_band_data(),
                    PCRs[9].get_band_data(),
                    PCRs[10].get_band_data()
                    // PCRs.map( () => o.get_band_data() )
                ]
                
                GEL_DATA = lanes_data.map(bands_data => { return bands_data.map(to_gel_band) })

                document.getElementById("status_div").innerHTML = "<b>Ready!</b>"

                return PCRs
            }

            download_data_as_file = function(data_rows, filename){
                let csvContent = "data:text/csv;charset=utf-8," + data_rows.map(e => e.join(",")).join("\n");

                var encodedUri = encodeURI(csvContent);
                var link = document.createElement("a");
                link.setAttribute("href", encodedUri);
                link.setAttribute("download", filename + ".csv");
                document.body.appendChild(link);

                link.click(); // This will download the data file named "my_data.csv".
            }

            get_fluorescence_data_by_cycle = function(){
                let fluorescence_data_rows = [ ["cycle"] ]
                let fluorescence_data_cols = []

                for (const [pcr_num, pcr_obj] of Object.entries(PCRs)){
                    fluorescence_data_rows[0].push(`PCR${pcr_num}`)
                    fluorescence_data_cols.push(pcr_obj.get_fluorescence_history())
                }

                for (let cycle_num=0; cycle_num<PCRs[1].cycles; cycle_num++){
                    my_row = [`${cycle_num}`]
                    for (let fdc of fluorescence_data_cols){
                        my_row.push(`${fdc[cycle_num]}`)
                    }
                    fluorescence_data_rows.push(my_row)
                }

                return fluorescence_data_rows
            }

            get_melting_curves = function(){ // !!! IN PROGRESS
                let T_array = []
                for (let T=30; T <=100; T += 0.5) T_array.push(T)

                let melting_curve_rows = [ ["T"] ]  // add a column for each PCR
                let melting_curve_cols = [ T_array ]

                for (const [pcr_num, pcr_obj] of Object.entries(PCRs)){
                    melting_curve_rows[0].push(`PCR${pcr_num}`)  // add header to first row
                    melting_curve_cols.push(pcr_obj.get_melting_curve(T_array))
                }

                // Collect rows across columns
                for (let i=0; i < T_array.length; i++){
                    let my_row = []
                    for (let col of melting_curve_cols){
                        my_row.push(col[i])
                    }
                    melting_curve_rows.push(my_row)
                }

                return melting_curve_rows
            }

        </script>
        <h1>qPCR Simulation App Demo</h1>

        <h3>Standard Primers</h3>
        <table>
            <tr><th>gene</th><th>Primer A</th><th>Primer B</th><th>Product size (bp)</th></tr>
            <tr><td>MX1</td><td>CAGCCTGCTGACATTGGGTA</td><td>CATTACTGGGGACCACCACC</td><td></td></tr>
            <tr><td>STAT1</td><td>CACCTAACGTGCTGTGCGTA</td><td>GGATGTTGGTGGAGGATTCT</td><td></td></tr>
            <tr><td>PSMB8</td><td>TTACCTGCTTGGCACCATGT</td><td>GCTGCCGACACTGAAATACG</td><td></td></tr>
            <tr><td>PSMB9</td><td>ACAGCCTTTTGCCATTGGTG</td><td>GCAATAGCGTCTGTGGTGAAG</td><td></td></tr>
            <tr><td>TAP1</td><td>CCAGTGGTCTGTTGACTCCC</td><td>AATGTCAGCCCCTGTAGCAC</td><td></td></tr>
            <tr><td>TAP2</td><td>CCTTGAACAATGTCGGCAGC</td><td>AAGTGCAGCACCCTCCTTAC</td><td></td></tr>
            <tr><td>IFNB1</td><td>CTTTGCTCTGGCACAACAGG</td><td>GTGGAGAAGCACAACAGGAGA</td><td></td></tr>
            <tr><td>IFIH1</td><td>TTGGACTCGGGAATTCGTGG</td><td>AACGATGGAGAGGGCAAGTC</td><td></td></tr>
            <tr><td>DDX58</td><td>AGAGCACTTGTGGACGCTTTA</td><td>TTGTTTTGCCACGTCCAGTC</td><td></td></tr>
            <tr><td>GAPDH</td><td>GCAAATTCCATGGCACCGTC</td><td>TCGCCCCACTTGATTTTGGA</td><td></td></tr>
        </table>

        <!--  STAT1 primer B was originally backwards: AGAATCCTCCACCAACATCC -->

        <hr />
        <p> Primers and reaction conditions you specify here will be run on all patient samples.</p>

        <form>
            Primer A: <textarea id="primerA" rows="1" cols="30">CAGCCTGCTGACATTGGGTA</textarea> <br />
            Primer B: <textarea id="primerB" rows="1" cols="30">CATTACTGGGGACCACCACC</textarea> <br />
            cycles: <input id="pcr_num_cycles" type="text" size="4" value="25" />
            annealing temperature: <input id="pcr_annealing_temp" type="text" size="4" value="55" />

            <br />
            <input type="button" value="run PCRs" onclick="go()" />
            <div id="status_div">&nbsp;</div>
            <br />
            <input type="button" value="fluorescence data by cycle" onclick="download_data_as_file(get_fluorescence_data_by_cycle(), 'fluorescence_data_per_cycle')" />
            <input type="button" value="melting curve" onclick="download_data_as_file(get_melting_curves(), 'qPCR_melting_curves')" />
            <input type="button" value="run gel" onclick="GEL.load_svg(GEL_DATA)" />
        </form>
        <div id="gel_display"></div>

        <hr/>
        <div id="gel2_display"></div>

<!--
TO DO:
* find the expected lengths of all products.
* find primers that don't work:
    STAT1 and DDX58 fail to produce any products.
        STAT1 is actually in fairly high concentration. If I take the reverse complement of STAT1 primer B ("GGATGTTGGTGGAGGATTCT") I get clean bands.
        DDX58 is in low concentration, but IFIH1 is similarly low concentration and we can detect that fine.
            Taking the reverse complement of primer B (GACTGGACGTGGCAAAACAA) didn't help.
            Neither did reverse complement of primer A: TAAAGCGTCCACAAGTGCTCT
            Neither did reversing both primers.
    IFNB1 produces extremely faint bands in patients 4 & 6-10; you can see them better if you plot the fluorescence data.
        (It makes strong bands if you change polymerase_survival_per_cycle = 1.00)
        This is the gene present in the lowest concentration by far (0 in the patients where we never see a band.)

Is this still true?: The program can become corrupted so it neads to be reloaded, I noticed this when trying out the primers.

pp = PCRs[1].potential_products[0]
pbs = pp.pbsA
seqA = pbs.primer_seq  // CAGCCTGCTGACATTGGGTA
// dH = pbsA.alignment.template_dH; dS = pbsA.alignment.template_dS

dH = -147200; dS = -393.3; primer_conc = 0.1


primer_conc=0.1; H=-145200; S=-389.79999999999995; T_celsius=55

PCRs[1].potential_products[0].seq.length
// MX1: 869
// STAT1: 174
// PSMB8: 307
// PSMB9: 388
// TAP1: 1471
// TAP2: 1127
// IFNB1: 112
// IFIH1: 543
// DDX58: ???
// GAPDH: 266

pp = PCRs[1].potential_products[0]
pp.fraction_template_bound(primer_conc, H, S, T_celsius)

// GAPDH: 'GCAAATTCCATGGCACCGTCCAGGGGCGGGCGCAGGCCGGATGTGTTCGCGCCGCTGCGGGCCGAGCCACATCGCTCAGACACCATGGGGAAGGTGAAGGTCGGAGTCAACGGATTTGGTCGTATTGGGCGCCTGGTCACCAGGGCTGCTTTTAACTCTGGTAAAGTGGATATTGTTGCCATCAATGACCCCTTCATTGACCTCAACTACATGGTTTACATGTTCCAATATGATTCCACCCATGGCTCGCCCCACTTGATTTTGGA'
// IFNB1: 'CTTTGCTCTGGCACAACAGGTTGCTCTGGCACAACAGGTAGTAGGCGACACTGTTCGTGTTGTCAACATGACCAACAAGTGTCTCCTCCAAGTGGAGAAGCACAACAGGAGA'
// STAT1: 'CACCTAACGTGCTGTGCGTAGGCTGCACCGCCGCGCCCCCGCCTAGCCCTTCCGGATCCTGCGCGCAGAAAAGTTTCATTTGCTGTATGCCATCCTCGAGAGCTGTCTAGGTTAACGTTCGCACTCTGTGTATATAACCTCGACAGTCTTGGCAGGATGTTGGTGGAGGATTCT'

pp = PCRs[1].potential_products[0]
pp.get_melting_curve()

-->


    </body>
</html>
